### ISSUE TITLE
volumeClaimTemplates for deployments

### ISSUE THREAD

**Issue by karolinepauls** [kubernetes/kubernetes#93019]
**What would you like to be added**:

Deployments also need volumeClaimTemplates.

**Why is this needed**:

Currently Kubernetes doesn't have a good story for deployments that need large-ish non-persistent local disk space. The default answer is "use hostPath", which is not a good one - it forces unnecessary inflation of node storage for workers which may not use it at the end of the day (until they scale up and they do).

Desired lifecycle: 
- pod is being started
- pod allocates storage (a PV or PVC)
- pod is running
- pod is being terminated
- storage is freed
- pod is terminated

StatefulSets, compared to deployments, are incredibly painful to use but they are the only workaround. With StatefulSets, there can be a cronjob that deletes statefulset PVCs in regular intervals. This however exposes race conditions in Kubernetes (https://github.com/kubernetes/kubernetes/issues/74374), so another cronjob is needed to clean up stuck pods

free worker volumes:
```
# Exploit the fact that PVC deletion is delayed until there are no users left.
# Skip "Pending" PVCs because these are not in use yet and if deleted, they'll
# make the worker pod stuck when initialising.
kubectl get pvc -l "release=$RELEASE_NAME,app=smartsolar-worker" \
            -ojsonpath='{range .items[*]}{.metadata.name}{" "}{.status.phase}{"\n"}{end}' \
  | grep -v 'Pending$' \
  | awk '{print $1}' \
  | xargs --no-run-if-empty -t -- kubectl delete pvc
```

```
kubectl get pod -l "release=$RELEASE_NAME,app=smartsolar-worker" \
                -ojsonpath='{range .items[*]}{.metadata.name}{" "}{.status.conditions[*].message}{"\n"}{end}' \
  | (grep 'persistentvolumeclaim ".*" not found' || true) \
  | awk '{ print $1 }' \
  | xargs -t --no-run-if-empty kubectl delete pod
```

**Comment by janetkuo** [comment]
With Deployments, there's no guarantee for Pod identity. Therefore, supporting `volumeClaimTemplates` for Deployments would be non-trivial. If there's a proposal for how this could be implemented, we'd be happy to take a look.

**Comment by karolinepauls** [comment]
When the pod is gone, storage is gone, the next pod will receive freshly allocated storage from the storage class. Identity is not really a problem, though there might be some need of garbage collection in case PV deletion on pod deletion unexpectably fails.

**Comment by winaterion** [comment]
I will also follow this request, it is the exact case I now encounter. I have a steteless app, but it need a large volume **on-demand** and not pre-mounted on a node which might never gets used. I couldn't find any solution on that unless I will make my app statefulset instead of deployment.

**Comment by mlycore** [comment]
> When the pod is gone, storage is gone, the next pod will receive freshly allocated storage from the storage class. Identity is not really a problem, though there might be some need of garbage collection in case PV deletion on pod deletion unexpectedly fails.

I think Pod identity is important. Since if there are 3 Pods shareing 1 PersistentVolume, this won't be a problem. But If 3 Pods mounted 3 difference PersistentVolume, how does a fresh Pod know which PersistentVolume shoud it be mounted?

**Comment by karolinepauls** [comment]
> I think Pod identity is important. Since if there are 3 Pods sharing 1 PersistentVolume, this won't be a problem. But if 3 Pods mounted 3 different PersistentVolumes, how does a fresh Pod know which PersistentVolume it should mount?

There is no sharing pod-scoped volumes (let's call them transient). One pod - one "transient volume". Pod kaputt - volume gone forever. New pod - new volume.

A pod wants to mount 3 "transient volumes" - OK. When it's gone - the volumes are gone.

**Comment by JarnoRFB** [comment]
> Currently Kubernetes doesn't have a good story for deployments that need large-ish non-persistent local disk space. The default answer is "use hostPath", which is not a good one - it forces unnecessary inflation of node storage for workers which may not use it at the end of the day (until they scale up).

I think that describes the use case and the problem very well. Two alternative solutions that I have found, would be

1. Automatic removal of PVCs created by StatefulSet https://github.com/kubernetes/enhancements/pull/1915 (currently a proposal)
2. Generic ephemeral volumes https://kubernetes.io/docs/concepts/storage/ephemeral-volumes/#generic-ephemeral-volumes (currently in alpha stage in kubernetes 1.19). I have not tried this feature, but it seems that it would provision an ephemeral volume for each pod and should probably also be usable in `Deployment`s.

I think it would be great if the use case described here could be directly addressed in kubernetes with a go to solution. Right now there seem to be a few different ideas of addressing this.

**Comment by immanuelfodor** [comment]
> 1. Automatic removal of PVCs created by StatefulSet kubernetes/enhancements#1915 (currently a proposal)

FYI, https://github.com/kubernetes/enhancements/pull/2440 took over and got merged recently.

**Comment by gregswift** [comment]
I hope this isnt a deviation to the point of this ticket (volumeClaimTemplates for deployments), but the in between state between "volumes that care about identity" and "volumes that are ephemeral with the host" is "volumes that just care that they are being used by the same deployment".

I was looking into rollout performance issues on a statefulset we have, and ran across #68397. So decided to look at hte problem from the opposite direction, why does my application need to be a statefulset? The answer, dedicated volumes per pod, and thus volumeClaimTemplates.

Say I have a set of workers processing a chunks of data.  They have no real identity requirement, but I do need isolated volumes per worker.  If one restarts, it will just process the data stored on its disk just like it was before.  Its identity isn't relevant to that work, because it reviews its defined work set and reports back on restart anyway.

We just recently proved this with our statefulset by renaming the statefulset and all of the volumes with no impact to our application's behavior.

Ultimately we could go either way (this issue or #68397 being solved), but I have to say that the way I just described our app sounds more like a Deployment than a StatefulSet.

**Comment by lllamnyp** [comment]
The described use-case (pod gone, volume gone forever, new pod - new volume) sounds like an emptyDir to me. How does it not satisfy you?

**Comment by JarnoRFB** [comment]
@lllamnyp As I understand it `emptyDir` goes either against the disk of the node or against RAM, but it is not possible to use external storage as with volumes. In cases where one needs a lot of ephemeral storage using the node's disk or RAM is not really an option.

**Comment by lllamnyp** [comment]
@JarnoRFB I see, this doesn't seem to be explicit in the issue. In this case ephemeral volumes seem to be the way to go, though I'm not sure if the deployment controller is capable of creating them on the fly.

**Comment by JarnoRFB** [comment]
@lllamnyp  I agree. I would prefer to use Deployments with on the fly provisioned scratch space for the described use case and reserve StatefulSet for applications where state is integral to the correct function of the application. I think Generic ephemeral volumes should be able to provide this.

**Comment by gregswift** [comment]
+1 to @JarnoRFB 's point about wanting to use node's disk or ram

Statefulsets are very focused on order and ensuring consistency of pod=disk.

For the use case I was describing I just need more specific storage than `emptyDir` provides, and to be able to hand that volume off to another pod if the pod has to go away.  I don't care about state, IP, name, etc for the pod.  Just that something takes over that data.  If I scale up, provision a new volume for those pods.

If I scale down, the PVC template would manage the behavior.

**Comment by lllamnyp** [comment]
/remove-lifecycle-rotten
@JarnoRFB does this solve your problem?
https://kubernetes-csi.github.io/docs/ephemeral-local-volumes.html

**Comment by JarnoRFB** [comment]
@lllamnyp Yes I guess the so called *generic ephemeral inline volumes* would indeed solve the issue, as the volume is ephemeral, but still provisioned through a storage class. Thank's for the pointer!

**Comment by gregswift** [comment]
I will say I dont believe that addresses what I added to the discussion, but am happy to separate it into a different issue.

**Comment by electropolis** [comment]
> When the pod is gone, storage is gone

That's not true if there is PVC created for PV and Retain is set.

**Comment by willzhang** [comment]
maybe openkruise support this：https://openkruise.io/docs/user-manuals/cloneset/#support-pvcs
